Mathlib/Computability/QueryComplexity/Basic.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2025 Tomaz Mascarenhas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Geoffrey Irving, Tomaz Mascarenhas
+-/
+import Mathlib.Algebra.Group.Basic
+import Mathlib.Algebra.Group.Nat.Defs
+import Mathlib.Computability.QueryComplexity.Defs
+import Mathlib.Tactic.Cases
+
+/-!
+## Basic properties of `Comp` and its Monad instance.
+-/
+
+open Classical
+noncomputable section
+
+variable {ι I : Type}
+variable {s t u : Set I}
+variable {α β γ : Type}
+
+namespace QueryComplexity
+
+namespace Comp
+
+/-- The definition of `Comp.map` -/
+lemma map_eq (f : α → β) (x : Comp ι s α) : f <$> x = x >>= (fun x ↦ pure (f x)) := rfl
+
+/-- `Comp` is a lawful monad -/
+instance : LawfulMonad (Comp ι s) := LawfulMonad.mk'
+  (id_map := by
+    intro α f
+    simp only [map_eq, id, bind, bind']
+    induction' f with _ _ _ _ _ _ h0 h1
+    · rfl
+    · simp only [bind', h0, h1])
+  (pure_bind := by intro α β x f; simp only [bind, bind'])
+  (bind_assoc := by
+    intro _ _ _ f _ _
+    simp only [bind]
+    induction' f with _ _ _ _ _ _ h0 h1
+    · rfl
+    · simp only [bind', h0, h1])
+
+/-- Running `pure` and `pure'` yields the original value -/
+@[simp]
+lemma value_pure' (x : α) (o : I → Oracle ι) : (pure' x : Comp ι s α).value o = x := by
+  simp only [value, run, map_pure]
+
+@[simp]
+lemma value_pure (x : α) (o : I → Oracle ι) : (pure x : Comp ι s α).value o = x := by
+  simp only [pure, value_pure']
+
+/-- The definition of `Comp.bind` as `simp` lemmas -/
+@[simp]
+lemma pure'_bind (x : α) (f : α → Comp ι s β) : (pure' x : Comp ι s α) >>= f = f x := rfl
+
+@[simp]
+lemma query'_bind (o : I) (m : o ∈ s) (y : ι) (f0 f1 : Comp ι s α)
+    (g : α → Comp ι s β) : query' o m y f0 f1 >>= g = .query' o m y (f0 >>= g) (f1 >>= g) := rfl
+
+/-- `pure` has cost 0 -/
+@[simp]
+lemma cost_pure (x : α) (o : I → Oracle ι) (i : I) : (pure x : Comp ι s α).cost o i = 0 := by
+  simp only [cost, run]
+
+/-- `pure` has cost' 0 -/
+@[simp]
+lemma cost'_pure (x : α) (o : Oracle ι) (i : I) : (pure x : Comp ι s α).cost' o i = 0 := by
+  simp only [cost', cost_pure]
+
+/-- `pure'` has cost 0 -/
+@[simp]
+lemma cost_pure' (x : α) (o : I → Oracle ι) (i : I) : (pure' x : Comp ι s α).cost o i = 0 := by
+  simp only [cost, run]
+
+/-- `pure'` has cost' 0 -/
+@[simp]
+lemma cost'_pure' (x : α) (o : Oracle ι) (i : I) : (pure' x : Comp ι s α).cost' o i = 0 := by
+  simp only [cost', cost_pure']
+
+/-- `query'` costs one query, plus the rest -/
+@[simp]
+lemma cost_query' {i : I} (m : i ∈ s) (y : ι) (f0 f1 : Comp ι s α) (o : I → Oracle ι) (j : I) :
+    (query' i m y f0 f1).cost o j =
+      (if j = i then 1 else 0) + (if (o i y) then f0.cost o j else f1.cost o j) := by
```

Comment:
No need for parentheses around `o i y` (but also this if will go away once you do the `Bool → Comp` thing).